packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -376,5 +427,55 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function hasValidReturnType(
+      services: ParserServicesWithTypeInformation,
+      node:
+        | TSESTree.FunctionExpression
+        | TSESTree.ArrowFunctionExpression
+        | TSESTree.FunctionDeclaration
+        | null,
+    ): boolean {
+      if (node != null) {
+        const checker = services.program.getTypeChecker();
+        const functionTSNode = services.esTreeNodeToTSNodeMap.get(node);
+        const functionType =
+          (ts.isFunctionExpression(functionTSNode) ||
+          ts.isArrowFunction(functionTSNode)
+            ? getContextualType(checker, functionTSNode)
+            : services.getTypeAtLocation(node)) ??
+          services.getTypeAtLocation(node);
```

Comment:
[Bug] Why do we get a contextual type for function expressions and arrow functions? 

The rule you took this code from has a detailed explanation of why it calls `getContextualType` for function expressions and arrow functions:

https://github.com/typescript-eslint/typescript-eslint/blob/f248e689b04bdfd81429977a731ba6b8ad82622e/packages/eslint-plugin/src/rules/no-unsafe-return.ts#L91-L99

This means that the following case is reported:

```ts
const test: () => any  = (): void => console.log()
```

(it should not be reported since `(): void => console.log()` is allowed with `ignoreVoidInVoid: true`)

[playground](https://deploy-preview-8632--typescript-eslint.netlify.app/play/#ts=5.4.3&fileType=.ts&code=MYewdgzgLgBFCm0BcMAUBKGBeAfDAhmAJ4zZrooBuIAlgCbZ6iQgA28AdKyAOYZA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90YANqRE0aB2iQANJkiVhnRQDVJUgJJNL01fjiIAvgF1wYFyBdA&tsconfig=&tokens=false)